### PR TITLE
Feature/add caching4 cost explorer results

### PR DESCRIPTION
**Pull request by damnski**

### CODE REVIEW

File: api/server.go [hunk 1]
```diff
@@ -42,9 +48,37 @@ func NewServer(config common.Config) error {
...
+			CacheExpireTime = 4 * time.Hour
+		} else {
+			CacheExpireTime = config.CacheExpireTime
+		}
+		if config.CachePurgeTime == nil {
+			CachePurgeTime = 4 * time.Hour
+		} else {
+			CachePurgeTime = config.CachePurgeTime
+		}
+	*/
```
Comment (fishnix) [inline api/server.go]: 👀

File: api/server.go [hunk 2]
```diff
@@ -11,20 +11,26 @@ import (
...
 
 	log "github.com/sirupsen/logrus"
 )
 
+var (
+	//Org will carry throughout the api and get tagged on resources
+	Org string
+
+	// StatusCaches is a map of in-memory caches for resource index status
+	StatusCaches = make(map[string]*cache.Cache)
```
Comment (fishnix) [inline api/server.go]: 👀 this isn't a status cache

Comment (fishnix) [reply api/server.go]: `ResultsCache` maybe...

File: api/server.go [hunk 3]
```diff
@@ -42,9 +48,37 @@ func NewServer(config common.Config) error {
...
+		if config.CachePurgeTime == nil {
+			CachePurgeTime = 4 * time.Hour
+		} else {
+			CachePurgeTime = config.CachePurgeTime
+		}
+	*/
 
 	// Create a shared Cost Explorer session
 	for name, c := range config.Accounts {
+		// Create a cache with a 4 hour default expiry and a 15 minute default purge time per provider
```
Comment (fishnix) [inline api/server.go]: ~~per provider~~

File: api/handlers_spaces.go [hunk 4]
```diff
@@ -105,11 +106,25 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
+		out, err = ceService.GetCostAndUsage(r.Context(), &input)
+		if err != nil {
+			msg := fmt.Sprintf("failed to get costs for space %s: %s", spaceID, err.Error())
+			handleError(w, errors.Wrap(err, msg))
+			return
+		}
+		StatusCaches[account].Set(spaceID, out, cache.DefaultExpiration)
+	} else {
+		out = c.([]*costexplorer.ResultByTime)
+		log.Debugf("found cached object: %s", out)
```
Comment (fishnix) [inline api/handlers_spaces.go]: this logic works and I think is what go-cache recommends... we could also cache the json `[]byte`.  we should keep an eye on memory utilization after we deploy to make sure everything is kosher.

File: api/server.go [hunk 5]
```diff
@@ -42,9 +48,37 @@ func NewServer(config common.Config) error {
...
+		log.Warn("Unexpected error with configured expireTime")
+		// set default expireTime
+		expireTime, _ = time.ParseDuration("4h")
+	}
+	purgeTime, err := time.ParseDuration(config.CachePurgeTime)
+	if err != nil {
+		log.Warn("Unexpected error with configured purgeTime")
+		//set default purgeTime
+		expireTime, _ = time.ParseDuration("15m")
+	}
```
Comment (fishnix) [inline api/server.go]: Lets check for `""` here first and consider it valid.  that way if you don't include it in the config, it doesn't throw a warning.  Then if you get an error from `ParseDuration` you should return (which will kill the API).

File: api/handlers_spaces.go [hunk 6]
```diff
@@ -105,11 +106,25 @@ func (s *server) SpaceGetHandler(w http.ResponseWriter, r *http.Request) {
...
-	if err != nil {
-		msg := fmt.Sprintf("failed to get costs for space %s: %s", spaceID, err.Error())
-		handleError(w, errors.Wrap(err, msg))
-		return
+	// Use the go-cache if the object is found cached,
+	// else get the object in the cost-api delivering it to the cache
+	var out []*costexplorer.ResultByTime
+	c, ok := StatusCaches[account].Get(spaceID)
+	if !ok || c == nil {
+		log.Debugf("cache empty for Org: %s, calling Cost-API", Org)
```
Comment (fishnix) [inline api/handlers_spaces.go]: add the `spaceID` to this log message?

**Comment by fishnix** [comment]
there's on other thought I had while I was playing with this.  although this is an authenticated endpoint, there's actually an attack vector since the cost explorer never returns a `404` where you could probably fill the cache with "empty" results.  we can talk about this later.

also, its so much faster to respond from cache!  🐎

File: api/server.go [hunk 7]
```diff
@@ -11,20 +11,26 @@ import (
 	"github.com/YaleSpinup/cost-api/costexplorer"
 	"github.com/gorilla/handlers"
 	"github.com/gorilla/mux"
+	cache "github.com/patrickmn/go-cache"
 
 	log "github.com/sirupsen/logrus"
 )
 
+var (
+	//Org will carry throughout the api and get tagged on resources
```
Comment (tenyo) [inline api/server.go]: 👀

File: config/config.example.json [hunk 8]
```diff
@@ -10,4 +10,6 @@
   "token": "xxxxxxxx-yyyy-zzzz-aaaa-bbbbbbbbbbb",
   "logLevel": "debug",
   "org": "localdev"
-}
```
Comment (tenyo) [inline config/config.example.json]: 👀Add comma at the end of the last line

File: api/server.go [hunk 9]
```diff
@@ -42,9 +48,37 @@ func NewServer(config common.Config) error {
...
+	expireTime, err := time.ParseDuration(config.CacheExpireTime)
+	if err != nil {
+		log.Warn("Unexpected error with configured expireTime")
+		// set default expireTime
+		expireTime, _ = time.ParseDuration("4h")
+	}
+	purgeTime, err := time.ParseDuration(config.CachePurgeTime)
+	if err != nil {
+		log.Warn("Unexpected error with configured purgeTime")
+		//set default purgeTime
```
Comment (tenyo) [inline api/server.go]: 👀

File: api/server.go [hunk 10]
```diff
@@ -11,20 +11,26 @@ import (
...
+	cache "github.com/patrickmn/go-cache"
 
 	log "github.com/sirupsen/logrus"
 )
 
+var (
+	//Org will carry throughout the api and get tagged on resources
+	Org string
+
+	// StatusCaches is a map of in-memory caches for resource index status
```
Comment (tenyo) [inline api/server.go]: 👀 index status

File: api/server.go (hunk 7 above)
Comment (damnski) [reply api/server.go]: What do you want to adjust?  I can't tell - thx

Comment (fishnix) [reply api/server.go]: put a space before `Org`, after the `//`

File: api/handlers_spaces.go (hunk 4 above)
Comment (damnski) [reply api/handlers_spaces.go]: did nothing more so far

**Comment by damnski** [comment]
thanks for the kind words :hamburger:

File: api/server.go [hunk 11]
```diff
@@ -43,8 +50,29 @@ func NewServer(config common.Config) error {
 	}
 	Org = config.Org
```
Comment (fishnix) [inline api/server.go]: do something like this instead below

```golang
	if config.CacheExpireTime == "" {
		// set default expireTime
		config.CacheExpireTim = "4h"
	}

	expireTime, err := time.ParseDuration(config.CacheExpireTime)
	if err != nil {
		log.Warn("Unexpected error with configured expiretime")
	}
```

File: config/config.example.json [hunk 12]
```diff
@@ -9,7 +9,7 @@
   },
   "token": "xxxxxxxx-yyyy-zzzz-aaaa-bbbbbbbbbbb",
   "logLevel": "debug",
-  "org": "localdev"
+  "org": "localdev",
   "cacheExpireTime": "4h",
   "cachePurgeTime": "15m"
-}
+}
```
Comment (fishnix) [inline config/config.example.json]: 👀

File: docker/config.deco.json [hunk 13]
```diff
@@ -14,5 +14,7 @@
   },
   "token": "{{ .api_token }}",
   "logLevel": "{{ .log_level }}",
-  "org": "{{ .spinup_org }}"
+  "org": "{{ .spinup_org }}",
+  "cacheExpireTime": "{{ .cache_expire_time }}",
+  "cachePurgeTime": "{{ .cache_purge_time }}"
 }
```
Comment (fishnix) [inline docker/config.deco.json]: 👀

**Pull request merged**